mmv1/templates/terraform/datasource_plural_fw.go.tmpl

```diff
@@ -0,0 +1,284 @@
+			},
+{{- end}}
+{{ if $.HasRegion -}}
+			"region": schema.StringAttribute{
+				Optional: true,
+				Computed: true,
+			},
+{{- end}}
+{{ if $.HasZone -}}
+			"zone": schema.StringAttribute{
+				Optional: true,
+				Computed: true,
+			},
+{{- end}}
+        },
+	}
+}
+
+func (r *{{$.ResourceName}}sDataSource) Read(ctx context.Context, req resource.ReadRequest, resp *resource.ReadResponse) {
+	var data {{$.ResourceName}}sModel
+	var metaData *fwmodels.ProviderMetaModel
+
+    // Read Provider meta into the meta model
+	resp.Diagnostics.Append(req.ProviderMeta.Get(ctx, &metaData)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Read Terraform configuration data into the model
+	resp.Diagnostics.Append(req.State.Get(ctx, &data)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	tflog.Trace(ctx, "read {{$.Name}} resource")
+
+    // read back {{$.Name}}
+	r.{{$.ResourceName}}sRefresh(ctx, &data, &resp.State, req, &resp.Diagnostics)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Save data into Terraform state
+	resp.Diagnostics.Append(resp.State.Set(ctx, &data)...)
+}
+
+func (r *{{$.ResourceName}}sDataSource) {{$.ResourceName}}FWRefresh(ctx context.Context, data *{{$.ResourceName}}FWModel, state *tfsdk.State, req interface{}, diag *diag.Diagnostics) {
+	var metaData *fwmodels.ProviderMetaModel
+    //load default values
+{{ if $.HasProject -}}
+    project := sDataSource.GetProjectFramework(data.Project, types.StringValue(r.providerConfig.Project), &resp.Diagnostics)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+{{- end }}
+{{ if $.HasRegion -}}
+	region := sDataSource.GetRegionFramework(data.Region, types.StringValue(r.providerConfig.Region), &resp.Diagnostics)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+{{- end }}
+{{ if $.HasZone -}}
+	zone := sDataSource.GetZoneFramework(data.Zone, types.StringValue(r.providerConfig.Zone), &resp.Diagnostics)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+{{- end }}
+
+	var schemaDefaultVals fwtransport.DefaultVars
+{{ if $.HasProject -}}
+	schemaDefaultVals.Project = project
+{{- end }}
+{{ if $.HasRegion -}}
+	schemaDefaultVals.Region = region
+{{- end }}
+{{ if $.HasZone -}}
+	schemaDefaultVals.Zone = zone
+{{- end }}
+
+	// Use provider_meta to set User-Agent
+	userAgent := fwtransport.GenerateFrameworkUserAgentString(metaData, r.providerConfig.UserAgent)
+
+	url := fwtransport.ReplaceVars(ctx, req, &resp.Diagnostics, schemaDefaultVals, r.providerConfig, "{{"{{"}}{{$.ProductMetadata.Name}}BasePath{{"}}"}}{{$.SelfLinkUri}}{{$.ReadQueryParams}}")
+    if resp.Diagnostics.HasError() {
+        return
+    }
+
+    log.Printf("[DEBUG] Refreshing {{ $.Name -}} data: %s", data.Id.ValueString())
+
+    headers := make(http.Header)
+{{- if $.CustomCode.PreRead }}
+    {{ $.CustomTemplate $.CustomCode.PreRead false -}}
+{{- end}}
+    res, err := fwtransport.SendRequest(fwtransport.SendRequestOptions{
+        Config: r.providerConfig,
+        Method: "LIST",
```

Comment:
i'm not sure if this would actually work in cases where the method isn't `LIST`

[`compute_disks` uses `GET`](https://cloud.google.com/compute/docs/reference/rest/v1/disks/list) instead just like all other data-sources, unless there's something else that i'm missing?

fwiw i was working on standardizing how we handle pagination within plural data-sources in this PR:
- https://github.com/GoogleCloudPlatform/magic-modules/pull/14687

though it's been a while since i've touched it due to other things coming up